Adopt the project with ViewPager2

ebraminio opened the pull request:
[ViewPager2](https://developer.android.com/jetpack/androidx/releases/viewpager2) is a recently added component to fix different issues of ViewPager which is itself built upon RecyclerView (though we haven't used this property of it here).

The specific thing that has motivated me to adopt the project with ViewPager2 is lack of RTL languages support on ViewPager which now is supported by ViewPager2, most annoying thing for me when I use the app with a device that uses Persian for its UI, see this for example,

![before](https://user-images.githubusercontent.com/833473/78783117-5aa61a80-79b8-11ea-99f4-f66233cf546c.gif)

This patch turns it into: (which is the correct scrolling for touch screens)

![after](https://user-images.githubusercontent.com/833473/78783136-61cd2880-79b8-11ea-8ec7-2d053bc6e1e2.gif)

So the scrolling won't be confusing for RTL languages users anymore. I've fixed minor things which will explain as comment for the patch.

Thanks!

ebraminio commented on app/build.gradle:
```diff
@@ -143,8 +143,9 @@ dependencies {
+    implementation "androidx.viewpager2:viewpager2:1.0.0"
     implementation "android.arch.work:work-runtime:$workManagerVersion"
-    implementation "com.google.android.material:material:1.0.0"
+    implementation "com.google.android.material:material:1.1.0"
```
Had to upgrade material version to use TabLayoutMediator which is added in 1.1.0.

ebraminio commented on app/src/main/java/de/danoeh/antennapod/view/PagerIndicatorView.java:
```diff
@@ -1,15 +1,17 @@
 import android.view.View;
+
 import androidx.annotation.Nullable;
-import androidx.vectordrawable.graphics.drawable.ArgbEvaluator;
```
It didn't build with the change and `androidx` one is apparently declared for internal use of the package.

ebraminio commented on app/src/main/java/de/danoeh/antennapod/view/PagerIndicatorView.java:
```diff
@@ -38,26 +40,29 @@ private void setup() {
             public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                 PagerIndicatorView.this.position = position + positionOffset;
+                if (ViewCompat.getLayoutDirection(pager) == ViewCompat.LAYOUT_DIRECTION_RTL) {
+                    PagerIndicatorView.this.position = numPages - 1 - PagerIndicatorView.this.position;
```
Flips the pager dots for RTL languages, it can't be done once for the view as direction isn't specified at that time and should be done when the view is attached.

ebraminio commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -1031,7 +1032,7 @@ public final Cursor getCompletedMediaCursor(int limit) {
         return db.query(TABLE_NAME_FEED_MEDIA, null,
                 KEY_PLAYBACK_COMPLETION_DATE + " > 0", null, null,
-                null, String.format("%s DESC LIMIT %d", KEY_PLAYBACK_COMPLETION_DATE, limit));
+                null, String.format(Locale.US, "%s DESC LIMIT %d", KEY_PLAYBACK_COMPLETION_DATE, limit));
```
It was causing

![image](https://user-images.githubusercontent.com/833473/78784096-f5ebbf80-79b9-11ea-92a8-04f550867e54.png)

A bug that was there before my changes but had to fix so can see the view.

ByteHamster commented:
Thank you very much for your pull request! It fixes many bugs that must have been there for a long time. 

It would be easier to merge if you separated the changes to multiple pull requests. For example, the SQL changes are independent from ViewPager2 and I would directly merge those without discussion ;)

Some things I noticed while skimming the code:

- You fixed some SQL problems in `PodDBAdapter`. Did you fix all cases where we build sql statements with numbers or just the ones that caused problems in your tests?
- In order to be able to swipe the media player up and down on each page, we had to add a hacky workaround for setting the scrolling child of the BottomSheetBehavior. This is probably still an issue, so [`ViewPagerBottomSheetBehavior`](https://github.com/AntennaPod/AntennaPod/blob/b5836650d18aa611b2fe2ca73d41ed58342c1e94/app/src/main/java/com/google/android/material/bottomsheet/ViewPagerBottomSheetBehavior.java) needs to be adapted to ViewPager2.
- Checkstyle fails because of the following lines:
    ```
    [ERROR] /home/circleci/AntennaPod/./app/src/main/java/de/danoeh/antennapod/fragment/DownloadsFragment.java:55: switch without "default" clause. [MissingSwitchDefault]
    [ERROR] /home/circleci/AntennaPod/./app/src/main/java/de/danoeh/antennapod/fragment/EpisodesFragment.java:59: switch without "default" clause. [MissingSwitchDefault]
    [ERROR] /home/circleci/AntennaPod/./app/src/main/java/de/danoeh/antennapod/fragment/preferences/StatisticsFragment.java:50: switch without "default" clause. [MissingSwitchDefault]
    ```
    Checkstyle only checks changed lines, so that new code follows the rules. You did not change the semantics but because you touched the lines, it complains. Would be great if you could fix the errors there.
- The release build test fails because of the update of `com.google.android.material`. That update adds so many more methods that AntennaPod no longer fits into a single dex file. I was trying to avoid updating that library. Is the `TabLayoutMediator` necessary to make it work? I am not experienced with ViewPager2, so this might be a stupid question ;)
  If it is really necessary, it looks like we have to switch to MultiDex (what I tried to avoid in the past).
- Regarding commit d938ed2 (Localize digits in queue episodes and parallel downloads numbers), we should probably stop appending text manually and use a placeholder in the string resource instead. This requires re-translation but I think this would be a lot more clean. This is not really related to the RTL changes, so feel free to skip this.
- Regarding the translation changes: We use [Transifex](https://www.transifex.com/antennapod/antennapod/) for translations. Changes to translated string files will be overridden when pulling from Transifex the next time. Merging the local changes with those from Transifex is a lot of work, so please only modify translations using Transifex.

ebraminio commented:
Thanks for the quick review! :)

> The release build test fails because of the update of com.google.android.material. That update adds so many more methods that AntennaPod no longer fits into a single dex file. I was trying to avoid updating that library. Is the TabLayoutMediator necessary to make it work? I am not experienced with ViewPager2, so this might be a stupid question ;)

Yes, or copying it into the project, guess we have to adopt that version anyway.

Will apply the rest. Just was in hurry to fix things I see while testing :)

ebraminio commented:
> You fixed some SQL problems in PodDBAdapter. Did you fix all cases where we build sql statements with numbers or just the ones that caused problems in your tests?

AFAIK all of them, will provide the patch separately.

> In order to be able to swipe the media player up and down on each page, we had to add a hacky workaround for setting the scrolling child of the BottomSheetBehavior. This is probably still an issue, so ViewPagerBottomSheetBehavior needs to be adapted to ViewPager2.

I am not sure where the hack is but viewpager2 has an advantage on being able to resize its height in a better way so guess need some help here.

> Checkstyle fails because of the following lines: 

Will apply.

>  If it is really necessary, it looks like we have to switch to MultiDex (what I tried to avoid in the past).

It is understandable that you are trying to avoid, not sure what to suggest :/

> Regarding commit d938ed2 (Localize digits in queue episodes and parallel downloads numbers), we should probably stop appending text manually and use a placeholder in the string resource instead. This requires re-translation but I think this would be a lot more clean. This is not really related to the RTL changes, so feel free to skip this.

Would you please handle this if won't take much time? Other I will do of course.

> Regarding the translation changes: We use Transifex for translations. Changes to translated string files will be overridden when pulling from Transifex the next time. Merging the local changes with those from Transifex is a lot of work, so please only modify translations using Transifex.

Started translation there. Thanks!

ByteHamster commented:
> I am not sure where the hack is but viewpager2 has an advantage on being able to resize its height in a better way so guess need some help here.

BottomSheetBehavior only allows one scrolling child but we use a ViewPager with multiple scrolling children (shownotes, chapters). So in order to support that, we subclassed `BottomSheetBehavior` using an androidx package name to access the package-private field that stores the scrolling child. Whenever the ViewPager is scrolled, we guess its currently active child view by assuming that `offscreenPageLimit` equals the total number of pages (otherwise, retrieving the currently visible child view requires reflection) and update the scrolling child in `BottomSheetBehavior`. This is pretty messy, unfortunately...

Have a look at ViewPagerBottomSheetBehavior. That class is pretty short and I hope that it can be adapted to ViewPager2 with only a few changes.

ebraminio commented:
Dropped the unrelated changes (and added your multidex check to my app :) https://github.com/persian-calendar/DroidPersianCalendar/commit/4859aa8144914f3a3b2d562efd1f6397c104b55f)

ebraminio commented:
Used a local copy of TabLayoutMediator instead upgrading material, guess should work now

ebraminio commented:
Bots are OK now! PTAL

ebraminio commented:
> Have a look at ViewPagerBottomSheetBehavior. That class is pretty short and I hope that it can be adapted to ViewPager2 with only a few changes.

It looks interesting, will investigate what more can be done about but in theory it should work now just like what was before, let's see.

ebraminio commented on app/src/main/java/de/danoeh/antennapod/fragment/AudioPlayerFragment.java:
```diff
@@ -127,11 +130,10 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
         // Required for getChildAt(int) in ViewPagerBottomSheetBehavior to return the correct page
-        pager.setOffscreenPageLimit(NUM_CONTENT_FRAGMENTS);
-        pager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
+        // pager.setOffscreenPageLimit(NUM_CONTENT_FRAGMENTS);
```
Done. https://github.com/AntennaPod/AntennaPod/pull/4020/commits/b30fae76b1f9fd4a4ccea53942513353e08acae0

Revealed a faulty check with the method documentation, https://developer.android.com/reference/androidx/viewpager2/widget/ViewPager2#setOffscreenPageLimit(int) it should  `Valid values are all values >= 1 *or* OFFSCREEN_PAGE_LIMIT_DEFAULT` instead and as the actual code also wants.

ebraminio replied:
Ok, already filed as https://issuetracker.google.com/issues/144669596 will apply their work around also

ebraminio commented on app/src/main/java/com/google/android/material/tabs/TabLayoutMediator.java:
```diff
@@ -0,0 +1,309 @@
+ * {@link #attach()} call. Changing the ViewPager2 or TabLayout will require a new instantiation of
+ * TabLayoutMediator.
+ */
+public final class TabLayoutMediator {
```
> Hmm. I want to keep the amount of code that we need to take care of as small as possible. At some point, we have to upgrade the dependencies. We already use 64,945 out of 65,536 methods, so we can not go much longer without multidex, anyway.

So what do you suggest? It isn't a fork, can (and should) be deleted after the upgrade.

ebraminio replied:
> I will play around with ProGuard to see if we can get a bit more room for new methods without enabling MultiDex. I mean, even big apps like WhatsApp work without MultiDex. Something seems to waste methods...

They may rely more on native codes I believe. I will also check what has consumed the methods number.

ByteHamster replied:
I brought it down to around 44k by removing `-keep class androidx.** { *; }` from the proguard config. Currently fine-tuning the proguard settings so that the tests pass.

ebraminio replied:
Let me know when you landed it. Thanks!

ByteHamster replied:
Done :) #4035 is merged, so we have a bit more room for library updates now

ebraminio commented on app/src/main/java/de/danoeh/antennapod/fragment/gpodnet/GpodnetMainFragment.java:
```diff
@@ -26,63 +35,59 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
-        private static final int NUM_PAGES = 2;
-        private static final int POS_TOPLIST = 0;
-        private static final int POS_TAGS = 1;
-        private static final int POS_SUGGESTIONS = 2;
```
Not in scope of this change but weird that NUM_PAGES is 2 but we have three pages here. Guess better to fix it later.

ebraminio replied:
@ByteHamster while this may take time please have a look if NUM_PAGES = 2 is really desired and nothing has got hide by mistake if possible thanks.

ByteHamster replied:
I created issue #4093 for this.

ebraminio commented:
PTAL

ebraminio commented:
Have updated trasifex with improved messages, wish to see the update soon also.

ebraminio commented on app/src/main/java/de/danoeh/antennapod/fragment/AudioPlayerFragment.java:
```diff
@@ -132,7 +132,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
         pager.setAdapter(new AudioPlayerPagerAdapter(this));
         // Required for getChildAt(int) in ViewPagerBottomSheetBehavior to return the correct page
-        // pager.setOffscreenPageLimit(NUM_CONTENT_FRAGMENTS);
+        pager.setOffscreenPageLimit((int) NUM_CONTENT_FRAGMENTS);
```
Applied another solution based on https://issuetracker.google.com/issues/144669596

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/ItemPagerFragment.java:
```diff
@@ -76,31 +77,21 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
         pager.setId(newId);
-        pager.setAdapter(new ItemPagerAdapter());
+        pager.setAdapter(new ItemPagerAdapter(this));
         pager.setCurrentItem(feedItemPos);
```
When clicking an episode (not the first one in the list), I now see an animation scrolling through the pages of other episodes. Not sure what to do about that.

ebraminio replied:
Done https://github.com/AntennaPod/AntennaPod/pull/4020/commits/e2ff176455db0f8fc43e0c1e8d92cdf80a316881

ebraminio commented on app/src/main/java/com/google/android/material/bottomsheet/ViewPagerBottomSheetBehavior.java:
```diff
@@ -31,8 +32,8 @@ View findScrollingChild(View view) {
-        if (view instanceof ViewPager) {
-            ViewPager viewPager = (ViewPager) view;
+        if (view instanceof ViewPager2) {
+            ViewPager2 viewPager = (ViewPager2) view;
             View currentViewPagerChild = viewPager.getChildAt(viewPager.getCurrentItem());
```
> With your changes, it is no longer possible to scroll up in the chapters list and the shownotes list (because every time when scrolling up, the slide-up sheet is closed).

I did understand the reasoning behind the code existence and have tested it locally, however I don't see a difference before these changes, assuming it is working as expected as up scrolling is still possible if one first goes for down scrolling then up scrolling. I will investigate more but can you check also if this really worked before my changes also?

> I am not really happy about getChildAt(0), so maybe you have another idea about that, too (IDs?).

Yes, that's not reliable. I used setBackground(Color.RED) to see if I've found the pages correctly.

ByteHamster replied:
> assuming it is working as expected as up scrolling is still possible if one first goes for down scrolling then up scrolling

Using the old version, I do not need to scroll up before being able to scroll down. I can just directly scroll up. In your version, I am also unable to hide the audio player by scrolling on the cover page.

> I used setBackground(Color.RED) to see if I've found the pages correctly.

When printing `currentViewPagerChild` in your current version, you can see that it always finds the RecyclerView that the ViewPager2 is based on. Not the Views within the pager.

ebraminio replied:
Oh, I am sorry, you already had provided the fix 🤦🤦‍♂️ https://github.com/AntennaPod/AntennaPod/pull/4020/commits/fca01a420ff267038a4d4b44e882cbf5c0b90667

ebraminio commented:
> The fragments store the last-used tab. With this PR, the indicator still shows the saved tab but the page that is actually displayed is always the first one. The displayed highlight is corrected after touching the screen for the first time.

Thanks for the reviews and putting screenshot which is really helpful. That's surprising… I don't have the issue in my use at https://github.com/persian-calendar/DroidPersianCalendar/blob/8e7259b15985219b8b9e42811d742a0bd980f40c/PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/CalendarFragment.kt#L174 (but we have simpler lifecycle there as we don't store/restore state there for onStop/onStart) should see why it happens here.

> While scrolling, the menu is sometimes duplicated (when showing that page for the first time). In previous versions, this was not the case.

I see… surprising! Should investigate.

> If you have shownotes with a big width (more than the screen width), you can no longer scroll their content horizontally.

More on https://github.com/AntennaPod/AntennaPod/pull/4020#discussion_r406841660

ebraminio commented:
> The fragments store the last-used tab. With this PR, the indicator still shows the saved tab but the page that is actually displayed is always the first one. The displayed highlight is corrected after touching the screen for the first time.

Fixed this one also I believe, had to move position restore logic to fix the issue.

ebraminio commented:
So guess this is the only remained one if your are ok with dfd6b9d 

> While scrolling, the menu is sometimes duplicated (when showing that page for the first time). In previous versions, this was not the case.

This one feels is out of my hand and rather a viewpager bug itself on when to call individual fragments' `onPrepareOptionsMenu`, but I think is a minor and negligible issue, do you like to revert viewpager2 for this page or do you have some clue what can I do about it? Thanks for all the helps btw :)

ByteHamster commented:
> So guess this is the only remained one

I think horizontal scrolling of shownotes is still broken

> do you like to revert viewpager2 for this page

This is an issue on all pages, so reverting on a single page would be not enough. I think it should rather be fixed

> but I think (in its current state) is a minor and negligible issue

I don't agree here. I think it is quite distracting to have the menu items jump and duplicate. Normally, you can move the pager without a lot of movement in the action bar. No idea how this can be fixed. We probably need to have a look at the ViewPager source code.

> Oh, I am sorry, you already had provided the fix

I am not totally happy about that fix, though. Feels quite hacky. If you have a better idea, I would love to change it.

ebraminio commented:
> I think horizontal scrolling of shownotes is still broken

Oh horizontally? Do you have a feed to test that? A solution to is described in, https://developer.android.com/training/animation/vp2-migration?hl=id#nested-scrollables

> This is an issue on all pages, so reverting on a single page would be not enough. I think it should rather be fixed

Only episodes view pager has options menu AFAIK?

> I am not totally happy about that fix, though. Feels quite hacky. If you have a better idea, I would love to change it.

Somewhere I needed something like this and just gave up and refactored to avoid that. Your solution however is much better than ones provided in https://stackoverflow.com/q/55728719 and https://stackoverflow.com/q/18609261 and https://stackoverflow.com/q/12854783 AFAICS

ByteHamster commented:
> Oh horizontally? Do you have a feed to test that?

Yes, horizontally. You can use the episode "Table shownotes" here: https://tools.bytehamster.com/podcast/rss.xml

While playing around with that, I noticed that the scrolling is already broken on the player page but it works when clicking an item. After your PR, it is broken in both places.

> Only episodes view pager has changing options menu AFAIK?

Downloads have a changing menu, too (empty - multi select - Clear log). Maybe the menu can be moved to the parent fragment that manages the ViewPager to prevent duplicate menus.

ebraminio commented:
Thank you for the feedback and suggestions, will investigate these, thanks :)

ebraminio commented:
> Downloads have a changing menu, too (empty - multi select - Clear log). Maybe the menu can be moved to the parent fragment that manages the ViewPager to prevent duplicate menus.

Done https://github.com/AntennaPod/AntennaPod/pull/4020/commits/b37b052d9113fec48169d76b2196a51f1ee9b07e works flawlessly AFAICS

ebraminio commented:
Fixed nested scroll of shownotes also https://github.com/AntennaPod/AntennaPod/pull/4020/commits/1e741c516d902898843b5da5e2eb07f4eb201a0c

ByteHamster commented on app/src/main/java/androidx/viewpager2/integration/NestedScrollableHost.java:
```diff
@@ -0,0 +1,133 @@
+        return super.onInterceptTouchEvent(e);
+    }
+
+    private void handleInterceptTouchEvent(MotionEvent e) {
```
Scrolling generally works fine now but it sometimes feels a bit strange.

- It is really easy to accidentally switch to another page when trying to scroll up/down. I think the sensitivity should be inverted: It should be easy to scroll up/down and harder to switch the page. Not sure how that would work, I have not looked at the code too closely yet.
- Sometimes, flinging (move finger fast, then lift and the view continues to scroll) does not work. This might be caused by the thing above.

ebraminio replied:
stuck with it, guess need your help :)

ebraminio replied:
wrapping it in NestedScrollView make it not scroll easily but causes more harm elsewhere, make it similar to bottomsheet shownote issue, I think ideally both webviews should be wrap content to make nestedscrollview handle their scrolling but that seems also tricky.

ByteHamster replied:
> I think ideally both webviews should be wrap content to make nestedscrollview handle their scrolling

This is not ideal. Wrapping in `nestedscrollview` causes me headache already because it increases the memory footprint of the WebView. Setting it to `wrap_content` makes it render the whole page, even the parts that are not visible. If the page contains many images, this can lead to crashes. So ideally, I want to go completely without the `nestedscrollview` hack.

> after playing some hours with it, no progress, guess need your help :)

Maybe implementing `NestedScrollingChild3` could be a starting point. I do not know much about that topic either, so I can not help too much, unfortunately.

ebraminio commented:
Reverted ViewPager2 adoption in ItemPagerFragment, I understand if you don't see this as a solution so will investigate how it can be fixed properly, but lack of RTL adoption isn't that issue here as there is not tablayout indicator to make things confusing.

ebraminio commented:
@ByteHamster Played some more time with it and now it looks good AFAICT, feel free to have a look to see if you feel that way. Thanks :)

ebraminio commented:
@ByteHamster (posting again, sorry for the spamming) Played some more time with it and now it looks good AFAICT, feel free to have a look to see if you also feel that way. Thanks :)

ebraminio commented:
> Looks good to me. Thanks a lot for all the work! Ready for merging?

I believe so. Thanks :)
